fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java

```diff
@@ -100,4 +120,59 @@ private void search(String source, String filePattern, MessageContext messageCon
             }
         }
     }
+
+    /**
+     *
+     * @param dir sub directory
+     * @param fileList list of file inside directory
+     * @param messageContext the message context that is generated for processing the file
+     */
+    private void getAllFiles(FileObject dir, List<FileObject> fileList, MessageContext
+            messageContext) {
+        try {
+            FileObject[] children = dir.getChildren();
+            for (FileObject child : children) {
+                fileList.add(child);
+            }
+        } catch (IOException e) {
+            handleException("Unable to list all folders", e, messageContext);
+        }
+    }
+
+    /**
+     *
+     * @param child sub folder
+     * @param filePattern pattern of the file to be searched
+     * @param messageContext the message context that is generated for processing the file
+     * @param factory OMFactory
+     * @param result OMElement
+     * @param ns OMNamespace
+     * @throws IOException
+     */
+    private void searchSubFolders(FileObject child, String filePattern, MessageContext
+            messageContext, OMFactory factory, OMElement result, OMNamespace ns) throws
+            IOException {
+        List<FileObject> fileList = new ArrayList<FileObject>();
+        getAllFiles(child, fileList, messageContext);
+        FilePattenMatcher fpm = new FilePattenMatcher(filePattern);
+        String outputResult;
+        try {
+            for (FileObject file : fileList) {
+                if (file.getType() == FileType.FILE) {
+                    if (fpm.validate(file.getName().getBaseName().toLowerCase())) {
+                        outputResult = file.getName().getPath();
+                        OMElement messageElement = factory.createOMElement(FileConstants.FILE,
+                                ns);
+                        messageElement.setText(outputResult);
+                        result.addChild(messageElement);
+                    }
+                } else if (file.getType() == FileType.FOLDER) {
+                    searchSubFolders(file, filePattern, messageContext, factory, result, ns);
+                }
+            }
+        } catch (IOException e) {
+            handleException("Unable to search a file in sub folder.", e, messageContext);
+        }
+    }
 }
```

Comment:
Don't you have to close file objects after using? Not sure please check this. Better do a load test for this feature.